File: misc/leetcode/unique-paths.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#define LL long long
class Solution {
   public:
    LL gcd(LL a, LL b) {
        while (true) {
            LL c = a % b;
            if (c == 0) {
                break;
            }
            a = b;
            b = c;
        }
        return b;
    }
    int uniquePaths(int m, int n) {
        LL a = 1, b = 1;
        for (int i = 1; i < n; i++) {
            a *= (LL)(m + i - 1);
            b *= (LL)i;
            LL c = gcd(a, b);
            a = a / c;
            b = b / c;
        }
        return (int)(a / b);
    }
};

File: leetcode/unique-paths.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        # C(n,m) = C(n-1,m-1) + C(n-1,m)
        N = n + m - 2
        M = m - 1
        dp = [[0] * (M + 1), [0] * (M + 1)]
        now = 0
        dp[now][0] = 1
        for _ in range(N):
            for x in range(M + 1):
                res = 0
                if x > 0:
                    res += dp[now][x - 1]
                res += dp[now][x]
                dp[1 - now][x] = res
            now = 1 - now
        return dp[now][M]


if __name__ == '__main__':
    s = Solution()
    print(s.uniquePaths(3, 2))
